File: c++/6-functions.cpp
#include <stdio.h>

int add(int a, int b)
{
    return a+b;
}

void echo(int e)
{
    printf("echo e=%d\n", e); 
    e--;
    if (e) echo(e);
}

int (*operation)(int a, int b);

void increment(int *what, int *to)
{
    printf("increment: received *what=%d what=%p *to=%d to=%p\n", *what, what, *to, to);
    (*what)++;
    *to++;
    printf("increment: returning *what=%d what=%p *to=%d, to=%p\n", *what, what, *to, to );
}

int main()
{
	int a=2, b=3;

	printf("%d + %d = %d\n", a, b, add(a, b));
	
	echo(10);
	
	operation = add;
	printf("result of operation on %d and %d = %d\n", a, b, operation(a, b));
	
	printf("a = %d b = %d\n", a, b);
	increment(&a, &b);
	printf("a after increment = %d\n", a);
	printf("b after increment = %d\n", b);
	return 0;
}

File: rust/6-functions.rs
fn add(a: i32, b: i32) -> i32 {
    a + b
}

fn echo(e: i32) {
    println!("echo e={}", e);
    if e > 1 {
        echo(e - 1);
    }
}

fn increment(what: &mut i32, to: &mut i32) {
    println!("increment: received *what={} *to={}", *what, *to);
    *what += 1;
    *to += 1;
    println!("increment: returning *what={} *to={}", *what, *to);
}

fn main() {
    let mut a = 2;
    let mut b = 3;

    println!("{} + {} = {}", a, b, add(a, b));

    echo(10);

    let operation = add;
    println!("result of operation on {} and {} = {}", a, b, operation(a, b));

    println!("a = {} b = {}", a, b);
    increment(&mut a, &mut b);
    println!("a after increment = {}", a);
    println!("b after increment = {}", b);
}